nextcloud_odoo_sync/models/nc_sync_user.py

```diff
@@ -1,103 +1,478 @@
+        if "nc_calendar_id" in vals:
+            calendar_event_ids = (
+                self.env["calendar.event"]
+                .search(
+                    [
+                        "|",
+                        ("user_id", "=", self.user_id.id),
+                        ("partner_ids", "in", self.user_id.partner_id.id),
+                        ("nc_synced", "=", False),
+                    ]
+                )
+                .filtered(
+                    lambda x: not x.nc_calendar_ids
+                    or (
+                        x.nc_calendar_ids
+                        and self.user_id.partner_id.id not in x.nc_calendar_ids.ids
+                    )
+                )
+            )
+            calendar_ids = calendar_event_ids.nc_calendar_ids.filtered(
+                lambda x: x.user_id == self.user_id
+            )
+            for calendar_id in calendar_ids:
+                calendar_event_ids.nc_calendar_ids = [(3, calendar_id.id)]
+            calendar_event_ids.with_context(sync=True).write(
+                {"nc_calendar_ids": [(4, vals["nc_calendar_id"])]}
+            )
         return super(NcSyncUser, self).write(vals)
 
+    def unlink(self):
+        """
+        Inherited odoo base function
+        :return add changes into this predefined functions
+        """
+        calendar_event_obj = self.env["calendar.event"]
+        sync_user_ids = self.search([]).mapped("user_id")
+        for record in self.filtered(lambda x: x.user_id):
+            calendar_ids = calendar_event_obj.search(
+                [("user_id", "=", record.user_id.id)]
+            )
+            for calendar in calendar_ids:
+                # Check if the calendar event is shared with another
+                # Odoo user with an active Nextcloud calendar sync
+                if len(calendar.nc_hash_ids.ids) > 1:
+                    hash_ids = calendar.nc_hash_ids.filtered(
+                        lambda x: x.user_id == record.user_id
+                        or (sync_user_ids and x.user_id not in sync_user_ids)
+                    )
+                    hash_ids.unlink()
+                    if not calendar.nc_hash_ids:
+                        calendar.write({"nc_uid": False, "nc_synced": False})
+                # Otherwise, remove both the nc_uid and hash_ids
+                else:
+                    calendar.write({"nc_uid": False, "nc_synced": False})
+                    calendar.nc_hash_ids.unlink()
+            # Remove all Nextcloud calendar records
+            record.user_id.nc_calendar_ids.unlink()
+        return super(NcSyncUser, self).unlink()
+
+    def save_user_config(self):
+        """
+        Returns calendar event action. Close the pop-up and display the
+        calendar event records
+        :return calendar event action
+        """
+        return self.env.ref("calendar.action_calendar_event").sudo().read()[0]
```

Comment:
There's a better way to do this, and also the `sudo()` would be not necessary anymore. The correct way is:

```suggestion
        return self.env["ir.actions.actions"]._for_xml_id("calendar.action_calendar_event")
```